### DISCUSSION TITLE
Javascript fetch post file to fastapi error

### DISCUSSION THREAD

**Discussion by vijaystroup** [fastapi/fastapi#7055] [Questions] [answered]
I'm having an issue recieving files from my frontend nextjs application to fastapi.

Here is my fastapi setup:
```python
from fastapi import FastAPI, UploadFile, File

app = FastAPI()

origins = [
    "http://localhost:3000"
]
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"]
)

@app.post('/upload')
async def upload_file(file: UploadFile = File(...)):
    print(file)

    return {'message': 'ok'}
```

Here is my frontend setup:
```js
  async function upload(e) {
    console.log(e.target.files[0])

    await fetch('http://localhost:8000/upload', {
      method: 'POST',
      headers: {
        'Content-Type': 'multipart/form-data'
      },
      body: e.target.files[0]
    })
  }

<input required type='file' onChange={e => upload(e)} />
```

Snapshot of the web console:
![image](https://user-images.githubusercontent.com/38572582/99888807-6976d000-2c1d-11eb-86ec-e6ad2683f329.png)

Error from fastapi:
`INFO:     127.0.0.1:54686 - "POST /upload HTTP/1.1" 400 Bad Request`

I'm not entirely sure what I'm doing wrong as it seems this is the correct procedure from the docs. I've also been tweaking at it a little, and sometimes I get a `INFO:     127.0.0.1:52251 - "POST /upload HTTP/1.1" 422 Unprocessable Entity` error from fastapi.

**Comment by rohansaw** [comment, marked as answer]
Hi,
You fastapi code should work :thinking: 
It think you should try uploading the file via Form data. Try using the following snippet in the frontend:
```
let data = new FormData()
data.append('file', e.target.files[0])

await fetch('http://localhost:8000/upload', {
  method: 'POST',
  body: data
}) 
```

**Comment by vijaystroup** [comment]
Okay, here are my different outcomes with changing only the frontend part of my initial code posted:

### 1
```js
async function upload(e) {
  await fetch('http://localhost:8000/upload', {
    method: 'POST',
    headers: {
      'Content-Type': 'multipart/form-data'
    },
    body: e.target.files[0]
  })
}
<input required type='file' onChange={e => upload(e)} />
```
`INFO:     127.0.0.1:61895 - "POST /upload HTTP/1.1" 400 Bad Request`

### 2
```js
async function upload(e) {
  await fetch('http://localhost:8000/upload', {
    method: 'POST',
    // headers: {
    //   'Content-Type': 'multipart/form-data'
    // },
    body: e.target.files[0]
  })
}
<input required type='file' onChange={e => upload(e)} />
```
`INFO:     127.0.0.1:61895 - "POST /upload HTTP/1.1" 422 Unprocessable Entity`

### 3
```js
async function upload(e) {
  await fetch('http://localhost:8000/upload', {
    let data = new FormData()
    data.append('file', e.target.files[0])
    method: 'POST',
    headers: {
      'Content-Type': 'multipart/form-data'
    },
    body: data
  })
}
<input required type='file' onChange={e => upload(e)} />
```
`INFO:     127.0.0.1:61895 - "POST /upload HTTP/1.1" 400 Bad Request`

### 4
```js
async function upload(e) {
  await fetch('http://localhost:8000/upload', {
    let data = new FormData()
    data.append('file', e.target.files[0])
    method: 'POST',
    // headers: {
    //   'Content-Type': 'multipart/form-data'
    // },
    body: data
  })
}
<input required type='file' onChange={e => upload(e)} />
```
`<starlette.datastructures.UploadFile object at 0x7faa02537070>`
`INFO:     127.0.0.1:62266 - "POST /upload HTTP/1.1" 200 OK`

Does anyone know the reasoning on why only the 4th method works whereas the other methods do not?

And just a side-note,
from what I've read, this isn't possible right?
```python
class Software(BaseModel):
    name: str
    desc: str
    url: str
    image: UploadFile = Image(...)
    image_path: str
```
Since with our post request we would have to do a `JSON.stringify()` on my whole object consisting of:
```js
  const [software, setSoftware] = useState({
    name: '',
    desc: '',
    url: '',
    image: {},
    image_path: ''
  })
```

I've seen some issues suggesting to do something like this:
```python
class Software(BaseModel):
    name: str
    desc: str
    url: str
    image_path: str

@app.post('/software')
async def post_software(software: Software, file: UploadFile= File(...)):
    """add a new software document"""

    print(file)

    # add software info to database
    software_ref = db.document(f'software/{software.name}')
    software_ref.set({
        'name': software.name,
        'description': software.desc,
        'url': software.url,
        'image_path': software.image_path
    })

    # add software image to storage
    # blob = bucket.blob(f'images/{software.image.__name__}')
    # await blob.upload_from_file(software.image)

    return {'software': software}
````
But I'm unsure how you would make a fetch post request with this type of acceptance since you can only pass one payload from my understanding.

**Comment by rohansaw** [comment]
In your frontend you can do create a params string from your object and then pass it as query params in addition to the form data.
This probably isn't get cleanest solution but it should do the job:

```
...
const params = new URLSearchParams({
    name: software.name,
            ...
    }).toString();

 await fetch('http://localhost:8000/upload?'+params, {
    let data = new FormData()
    data.append('file', e.target.files[0])
    method: 'POST',
    body: data
  })
...
```

With this you can send 
1) Your object without the file as stringified Queryparams
2) Your file via Formdata

and then in the backend create your model accordingly.

Hope this helps

**Comment by vijaystroup** [comment]
I solved this issue by initially setting the image state variable to `null` instead of `{}`.
When trying to make a new FormData(), it was accepting image as an empty object rather than a populated one.

Here is correct code:
```js
const [software, setSoftware] = useState({
    name: '',
    desc: '',
    url: '',
    image: null,
    image_path: ''
  })

submit(e) {
  let image = new FormData()
  image.append('file', software.image)
  await fetch('http://localhost:8000/software', {
        method: 'POST',
        body: image
    })
}
<input required id='image' type='file' onChange={e => { setSoftware({...software, image: e.target.files[0], image_path: 'images/'+e.target.files[0].name}) }

**Comment by tiangolo** [comment]
Thanks for the help here @rohansaw ! :clap: :bow: 

Thanks for reporting back and closing the issue @VijayStroup  :+1:

**Comment by es-brookfield** [comment]
@VijayStroup @rohansaw 

kings, you dropped these 👑 👑....
(thank you SO much for following up & closing - saved me some precious hours)

**Comment by minons1** [comment]
I also faced the same problem with `422 (Unprocessable Entitiy)` for sending FormData from my Javascript FrontEnd using `fetch()` API. My problem solved when I don't specify my `Content-Type : 'application/x-www-form-urlencoded'` in the header. Actually it is wierd beacause the docs says that it is the right content type when we deals with `Form()` request parameter. Hope this helps and there'll be improvement to solve this problem.
